File: sdk/reproit-ios/Sources/ReproIt/Causal.swift
import Foundation

/// Foundation URL loading adapter used only under a ReproIt run. It captures
/// through URLProtocol, logs the universal redacted exchange marker, and serves
/// exact capsule responses without touching the live backend.
final class ReproItCausalURLProtocol: URLProtocol {
    private static let lock = NSLock()
    private static var actionIndex = 0
    private static var ordinal = 0
    private static var used = Set<Int>()
    private static var exchanges: [[String: Any]] = []
    private static var excludePrefix: String?
    private static var installed = false
    private var loadingTask: URLSessionDataTask?

    static func install(excluding endpoint: String?) {
        guard !installed, ProcessInfo.processInfo.environment["REPROIT_CAUSAL"] == "1" else { return }
        installed = true
        excludePrefix = endpoint
        if let raw = ProcessInfo.processInfo.environment["REPROIT_CAPSULE_JSON"],
           let data = raw.data(using: .utf8),
           let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
            exchanges = root["exchanges"] as? [[String: Any]] ?? []
        }
        URLProtocol.registerClass(Self.self)
        let capabilities = "{\"http\":{\"status\":\"captured\"},\"http_replay\":{\"status\":\"captured\"}}"
        NSLog("REPROIT:CAPABILITIES %@", capabilities)
        mergeCapabilities(capabilities)
    }

    static func advanceAction() {
        lock.lock(); actionIndex += 1; ordinal = 0; lock.unlock()
    }

    override class func canInit(with request: URLRequest) -> Bool {
        guard installed, URLProtocol.property(forKey: "ReproItHandled", in: request) == nil,
              let scheme = request.url?.scheme?.lowercased(), scheme == "http" || scheme == "https" else { return false }
        if let prefix = excludePrefix, request.url?.absoluteString.hasPrefix(prefix) == true { return false }
        return true
    }

    override class func canonicalRequest(for request: URLRequest) -> URLRequest { request }

    override func startLoading() {
        guard let url = request.url else { return }
        Self.lock.lock()
        let action = Self.actionIndex
        let currentOrdinal = Self.ordinal
        Self.ordinal += 1
        let replaying = !Self.exchanges.isEmpty
        var matched: (Int, [String: Any])?
        if replaying {
            for (index, exchange) in Self.exchanges.enumerated() where !Self.used.contains(index) {
                if exchange["required"] as? Bool == true,
                   exchange["actor"] as? String == (ProcessInfo.processInfo.environment["REPROIT_DEVICE"] ?? "a"),
                   exchange["actionIndex"] as? Int == action,
                   (exchange["method"] as? String)?.uppercased() == (request.httpMethod ?? "GET").uppercased(),
                   Self.canonical(exchange["url"] as? String) == Self.canonical(url.absoluteString) {
                    matched = (index, exchange); break
                }
            }
        }
        if let (index, exchange) = matched {
            Self.used.insert(index)
            Self.lock.unlock()
            let status = exchange["status"] as? Int ?? 200
            let headers = exchange["responseHeaders"] as? [String: String] ?? [:]
            let body = exchange["responseBody"]
            let data: Data
            if let string = body as? String { data = Data(string.utf8) }
            else { data = (try? JSONSerialization.data(withJSONObject: body ?? "")) ?? Data() }
            let response = HTTPURLResponse(url: url, statusCode: status, httpVersion: "HTTP/1.1", headerFields: headers)!
            client?.urlProtocol(self, didReceive: response, cacheStoragePolicy: .notAllowed)
            client?.urlProtocol(self, didLoad: data)
            client?.urlProtocolDidFinishLoading(self)
            NSLog("CAPSULE:HIT %@", exchange["id"] as? String ?? "")
            return
        }
        Self.lock.unlock()
        if replaying {
            let message = "CAPSULE:MISS \(request.httpMethod ?? "GET") \(url.absoluteString) action=\(action)"
            NSLog("%@", message)
            client?.urlProtocol(self, didFailWithError: NSError(domain: "ReproItCapsule", code: 1, userInfo: [NSLocalizedDescriptionKey: message]))
            return
        }

        let mutable = (request as NSURLRequest).mutableCopy() as! NSMutableURLRequest
        URLProtocol.setProperty(true, forKey: "ReproItHandled", in: mutable)
        let config = URLSessionConfiguration.ephemeral
        config.protocolClasses = []
        loadingTask = URLSession(configuration: config).dataTask(with: mutable as URLRequest) { data, response, error in
            if let error { self.client?.urlProtocol(self, didFailWithError: error); return }
            guard let response = response as? HTTPURLResponse else { return }
            let data = data ?? Data()
            let requestHeaders = Self.redactHeaders(self.request.allHTTPHeaderFields ?? [:])
            let responseHeaders = Self.redactHeaders(response.allHeaderFields.reduce(into: [String: String]()) { $0[String(describing: $1.key)] = String(describing: $1.value) })
            let exchange: [String: Any] = [
                "id": "\(ProcessInfo.processInfo.environment["REPROIT_DEVICE"] ?? "a")-\(action)-\(currentOrdinal)",
                "actor": ProcessInfo.processInfo.environment["REPROIT_DEVICE"] ?? "a",
                "actionIndex": action, "ordinal": currentOrdinal, "protocol": url.scheme ?? "http",
                "method": self.request.httpMethod ?? "GET", "url": url.absoluteString,
                "requestHeaders": requestHeaders,
                "requestBody": Self.bodyValue(self.request.httpBody, headers: requestHeaders),
                "status": response.statusCode, "responseHeaders": responseHeaders,
                "responseBody": Self.bodyValue(data, headers: responseHeaders), "required": true,
            ]
            if let raw = try? JSONSerialization.data(withJSONObject: exchange), let line = String(data: raw, encoding: .utf8) {
                NSLog("REPROIT:EXCHANGE %@", line)
                Self.appendExchange(line)
            }
            self.client?.urlProtocol(self, didReceive: response, cacheStoragePolicy: .notAllowed)
            self.client?.urlProtocol(self, didLoad: data)
            self.client?.urlProtocolDidFinishLoading(self)
        }
        loadingTask?.resume()
    }

    override func stopLoading() { loadingTask?.cancel() }

    static func secret(_ key: String) -> Bool {
        let key = key.lowercased().filter { !"-_. ".contains($0) }
        return ["password", "passwd", "secret", "token", "authorization", "cookie", "email", "phone",
                "apikey", "publishablekey", "privatekey", "accesskey", "signingkey"].contains { key.contains($0) }
    }
    private static func appendExchange(_ line: String) {
        guard let path = ProcessInfo.processInfo.environment["REPROIT_NETWORK_FILE"] else { return }
        guard let data = (line + "\n").data(using: .utf8) else { return }
        if !FileManager.default.fileExists(atPath: path) { FileManager.default.createFile(atPath: path, contents: nil) }
        guard let handle = try? FileHandle(forWritingTo: URL(fileURLWithPath: path)) else { return }
        defer { try? handle.close() }
        do { try handle.seekToEnd(); try handle.write(contentsOf: data) } catch { }
    }
    private static func mergeCapabilities(_ raw: String) {
        guard let path = ProcessInfo.processInfo.environment["REPROIT_CAPABILITIES_FILE"],
              let update = raw.data(using: .utf8),
              let incoming = try? JSONSerialization.jsonObject(with: update) as? [String: Any] else { return }
        var existing: [String: Any] = [:]
        if let data = try? Data(contentsOf: URL(fileURLWithPath: path)),
           let value = try? JSONSerialization.jsonObject(with: data) as? [String: Any] { existing = value }
        incoming.forEach { existing[$0.key] = $0.value }
        if let data = try? JSONSerialization.data(withJSONObject: existing) { try? data.write(to: URL(fileURLWithPath: path), options: .atomic) }
    }
    private static func canonical(_ raw: String?) -> String {
        guard let raw, var components = URLComponents(string: raw) else { return raw ?? "" }
        components.scheme = components.scheme?.lowercased()
        components.host = components.host?.lowercased()
        if let items = components.queryItems {
            components.queryItems = items.sorted {
                if $0.name != $1.name { return $0.name < $1.name }
                return ($0.value ?? "") < ($1.value ?? "")
            }
        }
        return components.string ?? raw
    }
    private static func redactHeaders(_ headers: [String: String]) -> [String: String] {
        Dictionary(uniqueKeysWithValues: headers.map { ($0.key, secret($0.key) ? "<reproit:secret>" : $0.value) })
    }
    private static func bodyValue(_ data: Data?, headers: [String: String]) -> Any {
        guard let data, !data.isEmpty else { return NSNull() }
        if headers.first(where: { $0.key.lowercased() == "content-type" })?.value.contains("json") == true,
           let value = try? JSONSerialization.jsonObject(with: data) { return redact(value) }
        return "<reproit:body:length=\(data.count)>"
    }
    static func redact(_ value: Any) -> Any {
        if let values = value as? [Any] { return values.map(redact) }
        if let map = value as? [String: Any] {
            var result: [String: Any] = [:]
            for (key, child) in map {
                if secret(key) {
                    result[key] = child is String
                        ? "<reproit:string:length=\((child as! String).count)>"
                        : "<reproit:secret>"
                } else {
                    result[key] = redact(child)
                }
            }
            return result
        }
        return value
    }
}

File: sdk/reproit-tui-go/causal.go
package reproittui

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"sort"
	"strconv"
	"strings"
	"sync"
)

type causalExchange struct {
	ID              string            `json:"id"`
	Actor           string            `json:"actor"`
	ActionIndex     int               `json:"actionIndex"`
	Ordinal         int               `json:"ordinal"`
	Protocol        string            `json:"protocol"`
	Method          string            `json:"method"`
	URL             string            `json:"url"`
	RequestHeaders  map[string]string `json:"requestHeaders"`
	RequestBody     interface{}       `json:"requestBody,omitempty"`
	Status          int               `json:"status"`
	ResponseHeaders map[string]string `json:"responseHeaders"`
	ResponseBody    interface{}       `json:"responseBody,omitempty"`
	Required        bool              `json:"required"`
}

type causalRoundTripper struct {
	inner                           http.RoundTripper
	network, action, actor, exclude string
	replay                          []causalExchange
	mu                              sync.Mutex
	used                            map[int]bool
	prior, ordinal                  int
}

func secretKey(key string) bool {
	k := strings.ToLower(key)
	k = strings.NewReplacer("-", "", "_", "", ".", "", " ", "").Replace(k)
	sensitiveKeys := []string{
		"password", "passwd", "secret", "token", "authorization", "cookie", "email", "phone",
		"apikey", "publishablekey", "privatekey", "accesskey", "signingkey",
	}
	for _, s := range sensitiveKeys {
		if strings.Contains(k, s) {
			return true
		}
	}
	return false
}

func redactGo(value interface{}) interface{} {
	switch v := value.(type) {
	case map[string]interface{}:
		out := map[string]interface{}{}
		keys := make([]string, 0, len(v))
		for k := range v {
			keys = append(keys, k)
		}
		sort.Strings(keys)
		for _, k := range keys {
			if secretKey(k) {
				if s, ok := v[k].(string); ok {
					out[k] = fmt.Sprintf("<reproit:string:length=%d>", len([]rune(s)))
				} else {
					out[k] = "<reproit:secret>"
				}
			} else {
				out[k] = redactGo(v[k])
			}
		}
		return out
	case []interface{}:
		out := make([]interface{}, len(v))
		for i := range v {
			out[i] = redactGo(v[i])
		}
		return out
	default:
		return value
	}
}

func redactedHeaders(h http.Header) map[string]string {
	out := map[string]string{}
	for k, values := range h {
		if secretKey(k) {
			out[k] = "<reproit:secret>"
		} else {
			out[k] = strings.Join(values, ",")
		}
	}
	return out
}

func canonicalGoURL(raw string) string {
	u, err := url.Parse(raw)
	if err != nil {
		return raw
	}
	u.RawQuery = u.Query().Encode()
	return u.String()
}
func (c *causalRoundTripper) actionIndex() int {
	b, err := os.ReadFile(c.action)
	if err != nil {
		return 0
	}
	n, _ := strconv.Atoi(strings.TrimSpace(string(b)))
	return n
}

func (c *causalRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
	if c.exclude != "" && strings.HasPrefix(req.URL.String(), c.exclude) {
		return c.inner.RoundTrip(req)
	}
	c.mu.Lock()
	defer c.mu.Unlock()
	action := c.actionIndex()
	if action != c.prior {
		c.prior = action
		c.ordinal = 0
	}
	ordinal := c.ordinal
	c.ordinal++
	if c.replay != nil {
		for i, e := range c.replay {
			matches := !c.used[i] && e.Required && e.Actor == c.actor &&
				e.ActionIndex == action && strings.EqualFold(e.Method, req.Method) &&
				canonicalGoURL(e.URL) == canonicalGoURL(req.URL.String())
			if matches {
				c.used[i] = true
				body, _ := json.Marshal(e.ResponseBody)
				if s, ok := e.ResponseBody.(string); ok {
					body = []byte(s)
				}
				h := http.Header{}
				for k, v := range e.ResponseHeaders {
					h.Set(k, v)
				}
				return &http.Response{
					StatusCode: e.Status,
					Status:     fmt.Sprintf("%d replay", e.Status),
					Header:     h,
					Body:       io.NopCloser(bytes.NewReader(body)),
					Request:    req,
				}, nil
			}
		}
		return nil, fmt.Errorf("CAPSULE:MISS %s %s action=%d", req.Method, req.URL, action)
	}
	var requestBody interface{}
	if req.Body != nil {
		raw, _ := io.ReadAll(req.Body)
		req.Body.Close()
		req.Body = io.NopCloser(bytes.NewReader(raw))
		if json.Unmarshal(raw, &requestBody) == nil {
			requestBody = redactGo(requestBody)
		} else if len(raw) > 0 {
			requestBody = fmt.Sprintf("<reproit:body:length=%d>", len(raw))
		}
	}
	resp, err := c.inner.RoundTrip(req)
	if err != nil {
		return nil, err
	}
	raw, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	resp.Body.Close()
	resp.Body = io.NopCloser(bytes.NewReader(raw))
	var responseBody interface{}
	isJSON := strings.Contains(resp.Header.Get("content-type"), "json")
	if isJSON && json.Unmarshal(raw, &responseBody) == nil {
		responseBody = redactGo(responseBody)
	} else {
		responseBody = fmt.Sprintf("<reproit:body:length=%d>", len(raw))
	}
	e := causalExchange{
		ID:              fmt.Sprintf("%s-%d-%d", c.actor, action, ordinal),
		Actor:           c.actor,
		ActionIndex:     action,
		Ordinal:         ordinal,
		Protocol:        req.URL.Scheme,
		Method:          req.Method,
		URL:             canonicalGoURL(req.URL.String()),
		RequestHeaders:  redactedHeaders(req.Header),
		RequestBody:     requestBody,
		Status:          resp.StatusCode,
		ResponseHeaders: redactedHeaders(resp.Header),
		ResponseBody:    responseBody,
		Required:        true,
	}
	if line, err := json.Marshal(e); err == nil {
		if f, err := os.OpenFile(c.network, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0600); err == nil {
			fmt.Fprintln(f, string(line))
			f.Close()
		}
	}
	return resp, nil
}

// InstallCausalHTTP wraps http.DefaultTransport only while the Reproit runner
// has provisioned side files. It returns a restoration function.
func InstallCausalHTTP(excludePrefix string) func() {
	network, capsulePath := os.Getenv("REPROIT_NETWORK_FILE"), os.Getenv("REPROIT_CAPSULE")
	if network == "" && capsulePath == "" {
		return func() {}
	}
	var replay []causalExchange
	if capsulePath != "" {
		var cap struct {
			Exchanges []causalExchange `json:"exchanges"`
		}
		if raw, err := os.ReadFile(capsulePath); err == nil && json.Unmarshal(raw, &cap) == nil {
			replay = cap.Exchanges
		}
	}
	prior := http.DefaultTransport
	http.DefaultTransport = &causalRoundTripper{
		inner:   prior,
		network: network,
		action:  os.Getenv("REPROIT_ACTION_FILE"),
		actor:   envOr("REPROIT_DEVICE", "a"),
		exclude: excludePrefix,
		replay:  replay,
		used:    map[int]bool{},
	}
	mergeGoCapabilities(capsulePath != "")
	return func() { http.DefaultTransport = prior }
}

func envOr(key, fallback string) string {
	if value := os.Getenv(key); value != "" {
		return value
	}
	return fallback
}
func mergeGoCapabilities(replay bool) {
	path := os.Getenv("REPROIT_CAPABILITIES_FILE")
	if path == "" {
		return
	}
	value := map[string]interface{}{}
	if raw, err := os.ReadFile(path); err == nil {
		json.Unmarshal(raw, &value)
	}
	value["http"] = map[string]string{"status": "captured", "detail": "Go http.DefaultTransport"}
	value["http_replay"] = map[string]string{
		"status": "captured",
		"detail": "Go fail-closed RoundTripper",
	}
	raw, _ := json.Marshal(value)
	os.WriteFile(path, raw, 0600)
}
